### PR TITLE
WIP: Build BestEffortCallback.

### LINKED ISSUE #677: Plans for future of "spec-like" API
This issue documents our intentions for unifying some features of the "spec-like" API with the "standard" built-in plans and any user-defined plans. The motivation of this change is to avoid two disjoint forks of plans development.

For background, see [differences from non-spec-like plans](http://nsls-ii.github.io/bluesky/plans.html#differences-from-non-spec-like-plans).

## 1. Smarter Dispatcher

Motivation: Make it easier to apply nice subscriptions (e.g. plots) to all plans where they apply

Enhancement: Refactor the dependency-injection scheme in `bluesky.plans.spec_api` into a  "smart" Dispatcher-like object that can apply to *all* plans (including built-in standard plans, user-defined plans, and spec-like plans) auto-subscribing relevant subscriptions based on the contents of the RunStart document and/or EventDescriptors.

Planned Release: v0.9.0 (next release)

```python
# Currrent
RE(bluesky.spec_api.ct(...))  # provides table
RE(bluesky.plans.count(...))  # provides nothing

# Planned
RE(bluesky.spec_api.ct(...))  # provides table
RE(bluesky.plans.count(...))  # provides nothing
D = bluesky.default_dispatcher()
RE.subscribe('all', D)
RE(bluesky.plans.count(...))  # now provides table
```

`D` will contain logic for mapping metadata to appropriate subscriptions, such as 1D plots for 1D scans and various 2D plots for various types of 2D scans. It will provide methods to configure its behavior interactively, including an 'off' switch. It will absorb many of the attributes that we currently hang off of `gs`, such as `PLOT_Y` and `TABLE_COLS` (maybe). It does _not_ include `gs.DETS`. See below for the fate of `gs.DETS`.

## 2. Consistent Signature and Deprecating Global State Detectors

Motivation: In an early version, we introduced the global detectors list `gs.DETS` as a way to synchronize plans and automatic subscriptions. Now that plans report richer metadata, it's not necessary, and the role of `gs.DETS` can be replaced by a normal user-defined variable. This simplifies the bluesky codebase and the usage, and it allows users to switch between multiple different lists easily.

API break: Require the detector argument for spec-like plans.

Planned Release: v0.10.0

```python
# Current
gs.DETS = [detector1, detector2]
ct()  # implicitly uses gs.DETS for detectors

# Planned
d = [detector1, detector2]
ct(d)  # where `d` is a list of detectors, akin to how some users use `gs.DETS` now
# The old usage, `ct()` with no arguments, is an error!
```

### LINKED ISSUE #712: Add 'hints' key to plan metadata
All built-in plans will get a new metadata key like:

```python
def my_scan(detectors, motor, ...):
    # Just for a dumb example, this plan specifies that the first
    # two detectors should be plotted against the motor.
    # We are generating something in the spirit of the Vega Lite spec.
    # It can't be literal Vega Lite because it needs additional info
    # from the Event Descriptors to be complete it, so it's sort of a
    # Vega Lite template.
    md = {...,
          'hints': {'vis': [{'version': 0,
                            'data': 'primary',   # the name of the stream the data will come from
                            'mark': 'point',
                            'x': motor.name,
                            'y': [d.name for d in detectors[:2]]}]}
         }
```

This tells us which *devices* might be useful in a plot. The devices themselves add a `hints()` method -- see https://github.com/NSLS-II/ophyd/issues/413 -- that put into the Event Descriptors which *fields* a.k.a "data keys" from those devices should be used, such as whether we should plot against the motor readback or the motor setpoint.

These are all just hints designed to provide nice defaults for plotting tools, hence the name 'hints', designed so that we don't have to do complex guessing on the other end.

### PR DESCRIPTION
**Pull request by danielballan**
This closes #677 and also closes #712.

TO DO:

- [x] Add hints to all built-in plans (where applicable). Closes #712.
- [x] Incorporate PeakStats into the BestEffortCallback.
- [x] Plot monitored signals with the BestEffortCallback.
- [x] Plot 2D data with the BestEffortCallback.
- [x] Delete `global_state.py` and write detailed release notes on migration.
- [x] Delete `spec_api.py` and write detailed release notes on migration.
- [x] Write IPython magics `%ct`, `%ascan`, etc. as experimental shortcuts.
- [ ] Add hints to ophyd devices in a companion PR (not required for merging this one -- it will just make this one *better* when it's in).

### CODE REVIEW

File: bluesky/magics.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        if len(line.split()) % 2 != 0:
+            raise TypeError("Wrong parameters. Expected: "
+                            "%mov motor position (or several pairs like that)")
+        ns = Namespace(self.shell.user_ns)
+        RE = ns['RE']
+        args = []
+        strs = []
+        for motor, pos in partition(2, line.split()):
+            args.extend([ns[motor], le(pos)])
+            strs.extend([motor, repr(le(pos))])
```
Comment (tacaswell) [inline bluesky/magics.py]: do not use repr, just use original string againg

File: doc/source/api_changes.rst [hunk 2]
```diff
@@ -4,9 +4,181 @@ Release Notes
...
+
+        RE(ct())  # monitoring, flying, and baseline readings are added
+
+     And formerly, those settings only affected the behavior of the "SPEC-like"
+     plans, such as ``ct`` and ``ascan``. They were ignored by their
+     counterparts ``count`` and ``scan``, as well as user-defined plans. This
+     was not desirable!
+
+     This scheme has been replaced by the
+     :ref:`diagnostic preprocessor <diagnostic_preprocessor>`, which can be
```
Comment (tacaswell) [inline doc/source/api_changes.rst]: :bike: :houses: 'diagnostic preprocessor' sounds like a tool you reach for when things are broken, not day-to-day operations.

 - ~housekeeping~
 - standard
 - background
 - bookkeeping
 - moosefish
 - actuarial
 - status
 - health
 - checkup
 - tuneup

Comment (danielballan) [inline doc/source/api_changes.rst]: That name change would touch a lot of things unrelated to these release notes, so let's duke that out in a separate PR.

File: bluesky/magics.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        print('---> RE(mvr({}))'.format(', '.join(strs)))
+        plan = bp.mvr(*args)
+        return RE(plan)
+
+@magics_class
+class SPECMagics(Magics):
+
+    @line_magic
+    def ct(self, line):
+        if line.split():
```
Comment (danielballan) [inline bluesky/magics.py]: Yes. Agreed `strip` is clearer. Changed.

File: bluesky/plans.py [hunk 4]
```diff
@@ -2090,18 +2173,19 @@ def scan(detectors, motor, start, stop, num, *, per_step=None, md=None):
...
+           'num_intervals': num - 1,
+           'plan_args': {'detectors': list(map(repr, detectors)), 'num': num,
+                         'motor': repr(motor),
+                         'start': start, 'stop': stop,
+                         'per_step': repr(per_step)},
+           'plan_name': 'scan',
+           'plan_pattern': 'linspace',
+           'plan_pattern_module': 'numpy',
+           'plan_pattern_args': dict(start=start, stop=stop, num=num),
+           'hints': {'dimensions': [('primary', motor.hints()['fields'])]},
```
Comment (danielballan) [inline bluesky/plans.py]: Yes. Good catch.

**Pull request merged**